[css-aspect-ratio] The transferred min/max sizes should be constrained by defined sizes

cathiechen opened the pull request:
#### 94a7be7787d38909554344152db793c2c361eb17
<pre>
[css-aspect-ratio] The transferred min/max sizes should be constrained by defined sizes
<a href="https://bugs.webkit.org/show_bug.cgi?id=243321">https://bugs.webkit.org/show_bug.cgi?id=243321</a>

Reviewed by Rob Buis.

The aspect-ratio transferred min/max values should be capped/floored by the definite preferred or min/max sizes, see [1].
This patch added constrainLogicalMinMaxSizesByAspectRatio to handle this scenario. This function also handles the constraint
to the automatic content-based minimum sizes in the ratio-dependent axis which should be capped by the maximum size [2].

[1] <a href="https://www.w3.org/TR/css-sizing-4/#aspect-ratio-size-transfers">https://www.w3.org/TR/css-sizing-4/#aspect-ratio-size-transfers</a>
[2] <a href="https://www.w3.org/TR/css-sizing-4/#aspect-ratio-minimum">https://www.w3.org/TR/css-sizing-4/#aspect-ratio-minimum</a>

* LayoutTests/TestExpectations:
* Source/WebCore/rendering/RenderBox.cpp:
(WebCore::RenderBox::constrainLogicalMinMaxSizesByAspectRatio const):
(WebCore::RenderBox::constrainLogicalWidthInFragmentByMinMax const):
(WebCore::RenderBox::constrainLogicalHeightByMinMax const):
(WebCore::RenderBox::computeMinMaxLogicalWidthFromAspectRatio const):
(WebCore::RenderBox::computeMinMaxLogicalHeightFromAspectRatio const):
* Source/WebCore/rendering/RenderBox.h:

Canonical link: <a href="https://commits.webkit.org/253262@main">https://commits.webkit.org/253262@main</a>
</pre>

rwlbuis commented on LayoutTests/TestExpectations:
```diff
@@ -4628,12 +4628,6 @@ webkit.org/b/214462 imported/w3c/web-platform-tests/css/css-scoping/slotted-plac
-webkit.org/b/214463 imported/w3c/web-platform-tests/css/css-sizing/aspect-ratio/block-aspect-ratio-041.html [ ImageOnlyFailure ]
-webkit.org/b/214463 imported/w3c/web-platform-tests/css/css-sizing/aspect-ratio/block-aspect-ratio-042.html [ ImageOnlyFailure ]
-webkit.org/b/214463 imported/w3c/web-platform-tests/css/css-sizing/aspect-ratio/block-aspect-ratio-043.html [ ImageOnlyFailure ]
-webkit.org/b/214463 imported/w3c/web-platform-tests/css/css-sizing/aspect-ratio/block-aspect-ratio-045.html [ ImageOnlyFailure ]
```
There should be a select WPT test passing too.

rwlbuis replied:
Still needed :)

cathiechen replied:
Sorry, the select element WPT test failure was introduced by the old patch, the latest patch fixed it, so there is no changes in TestExpectations...

rwlbuis commented on Source/WebCore/rendering/RenderBox.cpp:
```diff
@@ -672,49 +672,101 @@ void RenderBox::applyTransform(TransformationMatrix& t, const RenderStyle& style
+
+    // Sizing constraints in either axis (the origin axis) should be transferred through the preferred aspect ratio. See https://www.w3.org/TR/css-sizing-4/#aspect-ratio-size-transfers
+    bool shouldCheckTransferredMinSize = dimension == ConstrainDimension::Width ? !styleToUse.logicalMinWidth().isSpecified() : !styleToUse.logicalMinHeight().isSpecified();
+    bool shouldCheckTransferredMaxSize = dimension == ConstrainDimension::Width ? !styleToUse.logicalMaxWidth().isSpecified() : !styleToUse.logicalMaxHeight().isSpecified();
```
I with this part would use less ternary operators but I did not see how so far.

rwlbuis commented on Source/WebCore/rendering/RenderBox.cpp:
```diff
@@ -672,49 +672,101 @@ void RenderBox::applyTransform(TransformationMatrix& t, const RenderStyle& style
-        return std::max(logicalHeight, computedLogicalHeight.value());
-    return logicalHeight;
+    std::optional<LayoutUnit> computedLogicalMinHeight = computeLogicalHeightUsing(MinSize, logicalMinHeight, intrinsicContentHeight);
+    LayoutUnit maxH = computedLogicalMaxHeight ? computedLogicalMaxHeight.value() : LayoutUnit::max();
```
Sorry I missed it, but we want no abbreviations, so maxHeight here.

cathiechen replied:
Done

rwlbuis commented on Source/WebCore/rendering/RenderBox.cpp:
```diff
@@ -672,49 +672,101 @@ void RenderBox::applyTransform(TransformationMatrix& t, const RenderStyle& style
-    return logicalHeight;
+    std::optional<LayoutUnit> computedLogicalMinHeight = computeLogicalHeightUsing(MinSize, logicalMinHeight, intrinsicContentHeight);
+    LayoutUnit maxH = computedLogicalMaxHeight ? computedLogicalMaxHeight.value() : LayoutUnit::max();
+    LayoutUnit minH = computedLogicalMinHeight ? computedLogicalMinHeight.value() : LayoutUnit();
```
Ditto.

cathiechen replied:
Done, thanks

rwlbuis commented on Source/WebCore/rendering/RenderBox.cpp:
```diff
@@ -5479,16 +5531,41 @@ LayoutUnit RenderBox::computeLogicalWidthFromAspectRatio(RenderFragmentContainer
+
+std::pair<LayoutUnit, LayoutUnit> RenderBox::computeMinMaxLogicalHeightFromAspectRatio() const
+{
+    const RenderStyle& styleToUse = style();
```
I think styleToUse is not needed, the getter should get optimized. Also this nicely will make the ASSERT the first statement.

rwlbuis commented on Source/WebCore/rendering/RenderBox.cpp:
```diff
@@ -5479,16 +5531,41 @@ LayoutUnit RenderBox::computeLogicalWidthFromAspectRatio(RenderFragmentContainer
 {
-    LayoutUnit blockMinSize = constrainLogicalHeightByMinMax(LayoutUnit(), std::nullopt);
-    LayoutUnit blockMaxSize = constrainLogicalHeightByMinMax(LayoutUnit::max(), std::nullopt);
+    const RenderStyle& styleToUse = style();
```
I think styleToUse is not needed, the getter should get optimized. Also this nicely will make the ASSERT the first statement.

rwlbuis commented on Source/WebCore/rendering/RenderBox.cpp:
```diff
@@ -5479,16 +5531,41 @@ LayoutUnit RenderBox::computeLogicalWidthFromAspectRatio(RenderFragmentContainer
+    LayoutUnit transferredMaxSize = LayoutUnit::max();
+    if (styleToUse.logicalMinHeight().isSpecified()) {
+        LayoutUnit blockMinSize = constrainLogicalHeightByMinMax(LayoutUnit(), std::nullopt);
+        if (blockMinSize > LayoutUnit())
```
You can combine these two statements.

rwlbuis commented on Source/WebCore/rendering/RenderBox.cpp:
```diff
@@ -5479,16 +5531,41 @@ LayoutUnit RenderBox::computeLogicalWidthFromAspectRatio(RenderFragmentContainer
+    }
+    if (styleToUse.logicalMaxHeight().isSpecified()) {
+        LayoutUnit blockMaxSize = constrainLogicalHeightByMinMax(LayoutUnit::max(), std::nullopt);
+        if (blockMaxSize != LayoutUnit::max())
```
Ditto.

rwlbuis commented on Source/WebCore/rendering/RenderBox.cpp:
```diff
@@ -5479,16 +5531,41 @@ LayoutUnit RenderBox::computeLogicalWidthFromAspectRatio(RenderFragmentContainer
+
+    if (styleToUse.logicalMinWidth().isSpecified()) {
+        LayoutUnit inlineMinSize = computeLogicalWidthInFragmentUsing(MinSize, styleToUse.logicalMinWidth(), availableWidth(), *containingBlock(), nullptr);
+        if (inlineMinSize > LayoutUnit())
```
Ditto.

rwlbuis commented on Source/WebCore/rendering/RenderBox.cpp:
```diff
@@ -5479,16 +5531,41 @@ LayoutUnit RenderBox::computeLogicalWidthFromAspectRatio(RenderFragmentContainer
+
+    if (styleToUse.logicalMaxWidth().isSpecified()) {
+        LayoutUnit inlineMaxSize = computeLogicalWidthInFragmentUsing(MaxSize, styleToUse.logicalMaxWidth(), availableWidth(), *containingBlock(), nullptr);
+        if (inlineMaxSize != LayoutUnit::max())
```
Ditto.

rwlbuis commented on Source/WebCore/rendering/RenderBox.cpp:
```diff
@@ -5479,16 +5531,41 @@ LayoutUnit RenderBox::computeLogicalWidthFromAspectRatio(RenderFragmentContainer
-        transferredMaxSize = inlineSizeFromAspectRatio(horizontalBorderAndPaddingExtent(), verticalBorderAndPaddingExtent(), style().logicalAspectRatio(), style().boxSizingForAspectRatio(), blockMaxSize);
-    // Minimum size wins over maximum size.
-    transferredMaxSize = std::max(transferredMaxSize, transferredMinSize);
+
```
This method has empty lines whereas computeMinMaxLogicalWidthFromAspectRatio has not. We want something consistent for these two methods which are quite similar.

rwlbuis commented on Source/WebCore/rendering/RenderBox.cpp:
```diff
@@ -672,49 +672,101 @@ void RenderBox::applyTransform(TransformationMatrix& t, const RenderStyle& style
     }
 
+    if (shouldCheckTransferredMinSize && transferredLogicalMinSize > LayoutUnit()) {
+        // The transferred min size should be capped by the definite maxmium size.
```
maximum*

rwlbuis commented on Source/WebCore/rendering/RenderBox.cpp:
```diff
@@ -672,49 +672,101 @@ void RenderBox::applyTransform(TransformationMatrix& t, const RenderStyle& style
+        constrainLogicalMinMaxSizesByAspectRatio(computedMinWidth, computedMaxWidth, logicalWidth, minimumSizeType, ConstrainDimension::Width);
+
+    logicalWidth = std::min(logicalWidth, computedMaxWidth);
+    return std::max(logicalWidth, computedMinWidth);
```
I think it can be considered to combine these two lines. But then it becomes a clamp I think?

cathiechen replied:
Yes, the difference is that  `std::clamp` handles the minimum value first, then the maximum value, that makes maximum value has higher priority than minimum value. We want the minimum value to have higher priority, so they shouldn't be combined to `std::clamp` here.

rwlbuis replied:
Thanks for explaining!

darinadler replied:
I ran into this same problem a while ago when trying to apply `std::clamp` on a lot more places in WebKit. I am kind of sad the the standard library got it backwards from what we almost always would find more useful.

rwlbuis commented on Source/WebCore/rendering/RenderBox.cpp:
```diff
@@ -672,49 +672,101 @@ void RenderBox::applyTransform(TransformationMatrix& t, const RenderStyle& style
+    if (styleToUse.hasAspectRatio())
+        constrainLogicalMinMaxSizesByAspectRatio(minHeight, maxHeight, logicalHeight, minimumSizeType, ConstrainDimension::Height);
+    logicalHeight = std::min(logicalHeight, maxHeight);
+    return std::max(logicalHeight, minHeight);
```
Ditto.

rwlbuis reviewed (commented):
Patch now LGTM!

cathiechen committed:
[css-aspect-ratio] The transferred min/max sizes should be constrained by defined sizes
https://bugs.webkit.org/show_bug.cgi?id=243321

Reviewed by Rob Buis.

The aspect-ratio transferred min/max values should be capped/floored by the definite preferred or min/max sizes, see [1].
This patch added constrainLogicalMinMaxSizesByAspectRatio to handle this scenario. This function also handles the constraint
to the automatic content-based minimum sizes in the ratio-dependent axis which should be capped by the maximum size [2].

[1] https://www.w3.org/TR/css-sizing-4/#aspect-ratio-size-transfers
[2] https://www.w3.org/TR/css-sizing-4/#aspect-ratio-minimum

* LayoutTests/TestExpectations:
* Source/WebCore/rendering/RenderBox.cpp:
(WebCore::RenderBox::constrainLogicalMinMaxSizesByAspectRatio const):
(WebCore::RenderBox::constrainLogicalWidthInFragmentByMinMax const):
(WebCore::RenderBox::constrainLogicalHeightByMinMax const):
(WebCore::RenderBox::computeMinMaxLogicalWidthFromAspectRatio const):
(WebCore::RenderBox::computeMinMaxLogicalHeightFromAspectRatio const):
* Source/WebCore/rendering/RenderBox.h:

Canonical link: https://commits.webkit.org/253262@main
